src/dynamic.cpp

```diff
@@ -333,6 +347,51 @@ void Dynamic::map_message_recursive(pTHX_ const Descriptor *descriptor, const st
     map_message(aTHX_ descriptor, perl_package, options);
 }
 
+std::string Dynamic::pbname_to_package(pTHX_ const std::string &pb_name, const std::string &perl_package_prefix) {
+    std::ostringstream oss;
+    oss << perl_package_prefix << "::";
+    size_t last = 0, next = 0;
+    while ((next = pb_name.find('.', last)) != string::npos) {
+        oss << pb_name.substr(last, next-last) << "::";
+        last = next + 1;
+    }
+    oss << pb_name.substr(last);
+    return oss.str();
+}
+
+void Dynamic::map_message_prefix_recursive(pTHX_ const Descriptor *descriptor, const string &perl_package_prefix, const MappingOptions &options) {
+	// avoid recursion loop
+	if (recursed_names.find(descriptor->full_name()) != recursed_names.end())
```

Comment:
Here, could you avoid the need for recursed_names by first checking descriptor_map and then calling map_message?

Reply:
Unfortunately that would cause a message to be skipped that had already been mapped, but whose internally used messages had not.  So, as it stands, you can do this (see proto from issue #27):

```perl
$dynamic->load_file( "a.proto" );
$dynamic->map( { message => 'a.A', to => 'PB' },
                 message => 'a.A', prefix => 'OtherPB' );
```

So that the message you want mapping is available at a known place, `PB` (as with `map_message()`), and any non-nested messages (that `map_message()` won't map) that it requires to be mapped are mapped somewhere or other under `OtherPB`.

If `descriptor_map` were checked, the second mapping would stop at `a.A` and never recurse in to it's use of `b.B`.

This was actually the scenario that prompted me to write the patch -- I wanted to be able to load an arbitrary message and populate it with data.  I required that any depended-upon message also be mapped, but without knowledge of what they were.

Reply:
If `map_message_prefix()` checked `descriptor_map`, you could actually still load an arbitrary message like this:

```perl
$dynamic->load_file( "a.proto" );
$dynamic->map( { message => 'a.A', prefix => 'PB' } );
```

But you'd have to access your message at `PB::a`, which means the developer would be required to manually parse the full message name and convert it to a package (e.g., convert `a.A` to `PB::a`) to be able to use it.

If this is preferable to you (rather than separately tracking which messages `map_message_prefix()` has visited), let me know and I'll make the changes.

Reply:
Added `108_message_prefix_mapping.t`